hack/generator/pkg/astmodel/package_reference.go

```diff
@@ -5,88 +5,30 @@
 
 package astmodel
 
-import (
-	"fmt"
-	"strings"
-
-	"github.com/pkg/errors"
-)
-
 const (
-	LocalPathPrefix       = "github.com/Azure/k8s-infra/hack/generated/apis/" // TODO: From config?
 	genRuntimePathPrefix  = "github.com/Azure/k8s-infra/hack/generated/pkg/genruntime"
 	GenRuntimePackageName = "genruntime"
 	GroupSuffix           = ".infra.azure.com"
 )
 
-var MetaV1PackageReference = MakePackageReference("k8s.io/apimachinery/pkg/apis/meta/v1")
-
-// PackageReference indicates which package a type belongs to
-type PackageReference struct {
-	packagePath string
-}
-
-// MakeLocalPackageReference Creates a new local package reference from a group and package name
-func MakeLocalPackageReference(groupName string, packageName string) PackageReference {
-	url := LocalPathPrefix + groupName + "/" + packageName
-	return PackageReference{packagePath: url}
-}
+var MetaV1PackageReference = MakeLibraryPackageReference("k8s.io/apimachinery/pkg/apis/meta/v1")
 
-// MakePackageReference creates a new package reference from a path
-func MakePackageReference(packagePath string) PackageReference {
-	return PackageReference{packagePath: packagePath}
+type PackageReference interface {
+	// IsLocalPackage returns a valud indicating whether this is a local package
+	IsLocalPackage() bool
```

Comment:
This doesn't feel quite right -- the interface here has knowledge that a specific implementer is going to be a `LocalPackage`, and that `Group` and `Version` are a thing?

Are there really two interfaces here?
`Package` and `LocalPackage` (or `GroupVersionPackage` or something) where `LocalPackage`/`GroupVersionPackage` has the `Group` and `Version` methods, and also embeds `Package` (which is the basic one)?

Then `IsLocalPackage` becomes `AsLocalPackage` that does a check + cast and returns a `LocalPackage`, where you can call the methods without fear of error (because you know you have the right type!)

Reply:
If you add `StoragePackage` that's fine it fits well into that model too as it'll just impl `LocalPackage` rather than `Package`

Reply:
I don't think there are different interfaces as both implementations are a `PackageReference` - just to packages that are found in different places.

Changing to `AsLocalPackage()` - or having consumers do the cast themselves - locks in the hierarchy and makes it difficult for other kinds of package reference to be added. 

Essentially, consumers shouldn't need to know about the various implementations; they should interact with solely the interface.

Reply:
But if you had written the interface without knowing what the implementers were going to be, you'd never have included a `Group` and a `Package` function (as that doesn't make sense in the context of a `Reference`). That suggests to me that we've leaked implementation details into our interface.

I'm suggesting a hierarchy where Package and LocalPackage are interfaces. LocalPackage is a Package (with some extra methods), so they would still both be `PackageReferences` and you could use anything which impls `LocalPackage` as a `Package` (since `LocalPackage` is a superset).

When you need those extra methods, yes you have to cast if you've only got `LocalPackage` - but that's what the current impls of `Group` and `Package` functions are doing anyway effectively (return an error for some impls, not for others).

Reply:
For example if I wanted to add a `GithubPackage` which was a package specifically from GitHub, and it had its own `GithubStarCount()` function, we wouldn't put that function on the interface and have all the other impls go implement that method and return an error from it would we?